lib/segment/src/index/field_index/binary_index.rs

```diff
@@ -0,0 +1,564 @@
+use std::sync::Arc;
+
+use bitflags::bitflags;
+use bitvec::prelude::*;
+use parking_lot::RwLock;
+use rocksdb::DB;
+
+use super::{CardinalityEstimation, PayloadFieldIndex, PrimaryCondition, ValueIndexer};
+use crate::common::rocksdb_wrapper::DatabaseColumnWrapper;
+use crate::entry::entry_point::{OperationError, OperationResult};
+use crate::telemetry::PayloadIndexTelemetry;
+use crate::types::{
+    FieldCondition, Match, MatchValue, PayloadKeyType, PointOffsetType, ValueVariants,
+};
+
+bitflags! {
+    /// Due to being able to store multi-values, the binary value is not a simple
+    /// boolean, but rather a pair of booleans, one for true value and one for false value.
+    pub struct BinaryItem: u8 {
+        const TRUE = 0b00000001;
+        const FALSE = 0b00000010;
+    }
+}
+
+impl BinaryItem {
+    pub fn from_bools(has_true: bool, has_false: bool) -> Self {
+        let mut item = Self::empty();
+        item.set(Self::TRUE, has_true);
+        item.set(Self::FALSE, has_false);
+        item
+    }
+}
+
+impl From<bool> for BinaryItem {
+    fn from(src: bool) -> Self {
+        if src {
+            BinaryItem::TRUE
+        } else {
+            BinaryItem::FALSE
+        }
+    }
+}
+
+struct BinaryMemory {
+    trues: BitVec,
+    falses: BitVec,
+    trues_count: usize,
+    falses_count: usize,
+    indexed_count: usize,
+}
+
+impl BinaryMemory {
+    pub fn new() -> Self {
+        Self {
+            trues: BitVec::new(),
+            falses: BitVec::new(),
+            trues_count: 0,
+            falses_count: 0,
+            indexed_count: 0,
+        }
+    }
+
+    pub fn get(&self, id: PointOffsetType) -> BinaryItem {
+        debug_assert!(self.trues.len() == self.falses.len());
+        if (id as usize) >= self.trues.len() {
+            return BinaryItem::empty();
+        }
+
+        unsafe {
```

Comment:
Unsafe functions look scary, why not just use get instead?

Reply:
This skips a bound check on the second bitvec, which we can expect to be the same length.

But I agree that this is likely so insignificant that doing it without `unsafe` is better.